File: src/beagle/compiler/generator/ansic/StructPrinter.java
package beagle.compiler.generator.ansic;


public class StructPrinter extends CodePrinter
{

	public StructPrinter( StringBuilder buffer )
	{
		super(buffer);
	}

	public StructPrinter( StringBuilder buffer, int tabSize )
	{
		super(buffer, tabSize);
	}

	public void open( int modifiers, String name )
	{
		indent();
		if ((modifiers & Constants.PRIVATE) > 0)
			buffer.append("static ");
		if (name == null)
			buffer.append("struct {\n");
		else
		{
			buffer.append("struct ");
			buffer.append(name);
			buffer.append(" {\n");
		}
		incIndent();
	}

	public void open(  String name )
	{
		open(Constants.PRIVATE, name);
	}
	
	public void close()
	{
		decIndent();
		indent();
		buffer.append("};\n\n");
	}

	public void addPrimitive( int modifiers, String type, String name )
	{
		indent();
		if ((modifiers & Constants.CONST) > 0) buffer.append("const ");
		if ((modifiers & Constants.STRUCT) > 0) buffer.append("struct ");
		buffer.append(type);
		if ((modifiers & Constants.POINTER) > 0) buffer.append("*");
		buffer.append(" ");
		buffer.append(name);
		buffer.append(";\n");
	}

	public void addStruct( int modifiers, String type, String name )
	{
		addPrimitive(modifiers | Constants.STRUCT, type, name);
	}

}

File: modules/compiler/source/StructPrinter.cc
#include "StructPrinter.hh"
#include "CodePrinter.hh"


namespace beagle {
namespace compiler {


StructPrinter::StructPrinter(
    int tabSize ) : CodePrinter(tabSize)
{
    // nothing to do
}


StructPrinter::StructPrinter(
    CodePrinter &parent,
    int tabSize ) : CodePrinter(parent, tabSize)
{
    // nothing to do
}

void StructPrinter::open(
    int modifiers,
    const std::string &name )
{
    indent();
    getStream() << "struct " << name << "\n";
    indent();
    getStream() << "{\n";
    incIndent();
}

void StructPrinter::open(
    const std::string & name )
{
    open(0, name);
}

void StructPrinter::close()
{
    decIndent();
    indent();
    getStream() << "};\n\n";
}

void StructPrinter::addPrimitive(
    int modifiers,
    const std::string &type,
    const std::string &name )
{
    indent();
    if ((modifiers & CONST) > 0)
        getStream() << "const ";
    if ((modifiers & STRUCT) > 0)
        getStream() << "struct ";
    getStream() << type;
    if ((modifiers & POINTER) > 0)
        getStream() << "*";
    if ((modifiers & ARRAY) > 0)
        getStream() << "*";
    getStream() << " " << name << ";" << std::endl;
}

void StructPrinter::addStruct(
    int modifiers,
    const std::string & type,
    const std::string & name )
{
    addPrimitive(modifiers | STRUCT, type, name);
}


} // namespace compiler
} // namespace beagle
